File: LeetCode/Graph/DFS/JS/Leetcode-successor-lcci.js
var inorderSuccessor = function (root, p) {
    let [stk, cur, pre] = [[], root, null];
    while (stk.length || cur) {
        while (cur) {
            stk.push(cur);
            cur = cur.left;
        }
        cur = stk.pop();
        if (pre === p) return cur;
        pre = cur;
        cur = cur.right;
    }
    return null;
};

File: LeetCode/Graph/DFS/Python/Leetcode-successor-lcci.py
class Solution:
    def inorderSuccessor(self, root: TreeNode, p: TreeNode) -> TreeNode:
        stk, cur, pre = [], root, None
        while stk or cur:
            while cur:
                stk.append(cur)
                cur = cur.left
            cur = stk.pop()
            if pre == p:
                return cur
            pre = cur
            cur = cur.right
        return None
